File: config_loader.py
"""Load the shared gridded-pipeline configuration.

``config.yml`` beside this module is the canonical set of defaults.  A file
named by ``DSSAT_CONFIG_FILE`` is a *partial override* merged over those
defaults.  This keeps study-specific YAML files small without falling back to
different hidden defaults in the R and Python pipelines.

This module is mirrored by :mod:`config_loader.R`; keep their precedence and
blank-value behavior aligned.
"""
import os

_CONFIG = {}


def _deep_merge(base, override):
    """Return a recursive mapping merge without mutating either input."""
    merged = dict(base)
    for key, value in override.items():
        if isinstance(merged.get(key), dict) and isinstance(value, dict):
            merged[key] = _deep_merge(merged[key], value)
        else:
            merged[key] = value
    return merged


def _read_yaml(path, yaml):
    with open(path, "r", encoding="utf-8-sig") as fh:
        cfg = yaml.safe_load(fh) or {}
    if not isinstance(cfg, dict):
        raise ValueError(f"Top level must be a YAML mapping: {path}")
    return {str(k).lstrip("\ufeff"): v for k, v in cfg.items()}


def _validate_override(base, override, prefix=""):
    for key, value in override.items():
        path = f"{prefix}.{key}" if prefix else key
        if key not in base:
            raise ValueError(f"Unknown configuration key: {path}")
        expected = base[key]
        if isinstance(expected, dict) and isinstance(value, dict):
            _validate_override(expected, value, path)
        elif value is not None and value != "" and expected is not None:
            if isinstance(expected, bool) and not isinstance(value, bool):
                raise TypeError(f"{path} must be boolean, got {type(value).__name__}")
            if isinstance(expected, (int, float)) and not isinstance(expected, bool) \
                    and not isinstance(value, (int, float)):
                raise TypeError(f"{path} must be numeric, got {type(value).__name__}")
            if isinstance(expected, list) and not isinstance(value, list):
                raise TypeError(f"{path} must be a list, got {type(value).__name__}")


def _find_and_load():
    here = os.path.dirname(os.path.abspath(__file__))
    base_path = os.path.join(here, "config.yml")
    env_path = os.environ.get("DSSAT_CONFIG_FILE", "")
    try:
        import yaml  # PyYAML
    except ImportError as exc:
        raise RuntimeError(
            "PyYAML is required to load the central config.yml; install the "
            "project requirements before running the pipeline."
        ) from exc
    if not os.path.isfile(base_path):
        raise FileNotFoundError(f"Canonical pipeline config not found: {base_path}")
    if env_path and not os.path.isfile(env_path):
        raise FileNotFoundError(
            f"DSSAT_CONFIG_FILE points to a missing file: {env_path}"
        )
    try:
        cfg = _read_yaml(base_path, yaml)
        loaded = [base_path]
        if env_path and os.path.abspath(env_path) != os.path.abspath(base_path):
            override = _read_yaml(env_path, yaml)
            _validate_override(cfg, override)
            cfg = _deep_merge(cfg, override)
            loaded.append(env_path)
        print(
            f"[config_loader] Loaded {len(cfg)} settings from "
            + " + ".join(loaded)
        )
        return cfg
    except Exception as exc:  # noqa: BLE001
        raise RuntimeError(f"Failed to load pipeline configuration: {exc}") from exc


_CONFIG = _find_and_load()


def cfg_get(key, default):
    """Return config.yml[key], or `default` if absent or blank ("")."""
    if key not in _CONFIG or _CONFIG[key] is None:
        return default
    val = _CONFIG[key]
    if isinstance(val, str) and val == "":   # blank => default
        return default
    return val

File: config_loader.R
# config_loader.R
# ---------------------------------------------------------------------------
# Loader for the shared config.yml.
#
# Sourced near the top of SECTION 0 in dssat_main_pipeline.R. Exposes one
# config.yml beside this file is the canonical default mapping. A file named by
# DSSAT_CONFIG_FILE is merged over it as a partial study-specific override.
# This is the R twin of config_loader.py; keep precedence and blank handling
# aligned.
# ---------------------------------------------------------------------------

.CONFIG <- list()

local({
  deep_merge <- function(base, override) {
    if (!is.list(base)) base <- list()
    if (!is.list(override)) return(base)
    for (key in names(override)) {
      if (is.list(base[[key]]) && is.list(override[[key]])) {
        base[[key]] <- deep_merge(base[[key]], override[[key]])
      } else {
        base[[key]] <- override[[key]]
      }
    }
    base
  }

  validate_overlay <- function(base, override, prefix = "") {
    for (key in names(override)) {
      path <- if (nzchar(prefix)) paste(prefix, key, sep = ".") else key
      if (!(key %in% names(base))) stop("Unknown configuration key: ", path, call. = FALSE)
      if (is.list(override[[key]]) && is.list(base[[key]]) &&
          !is.null(names(override[[key]]))) {
        validate_overlay(base[[key]], override[[key]], path)
      } else if (is.list(override[[key]]) != is.list(base[[key]]) &&
                 length(override[[key]]) != 0L) {
        stop("Configuration type mismatch at ", path, call. = FALSE)
      }
    }
  }

  env_path <- Sys.getenv("DSSAT_CONFIG_FILE", "")
  this_dir <- tryCatch({
    if (requireNamespace("this.path", quietly = TRUE)) dirname(this.path::this.path()) else NA
  }, error = function(e) NA)
  if (is.na(this_dir)) this_dir <- getwd()
  base_path <- file.path(this_dir, "config.yml")
  if (!file.exists(base_path)) stop("Canonical pipeline config not found: ", base_path)
  if (nzchar(env_path) && !file.exists(env_path)) {
    stop("DSSAT_CONFIG_FILE points to a missing file: ", env_path)
  }
  if (!requireNamespace("yaml", quietly = TRUE)) {
    stop("Package 'yaml' is required to load the central config.yml.")
  }
  cfg <- yaml::read_yaml(base_path)
  if (!is.list(cfg)) stop("Top level must be a YAML mapping: ", base_path)
  loaded <- base_path
  if (nzchar(env_path) && normalizePath(env_path) != normalizePath(base_path)) {
    override <- yaml::read_yaml(env_path)
    if (!is.list(override)) stop("Top level must be a YAML mapping: ", env_path)
    validate_overlay(cfg, override)
    cfg <- deep_merge(cfg, override)
    loaded <- c(loaded, env_path)
  }
  assign(".CONFIG", cfg, envir = .GlobalEnv)
  message(sprintf("[config_loader] Loaded %d settings from %s",
                  length(cfg), paste(loaded, collapse = " + ")))
})

# cfg_get("key", default):
#   - returns default when key is absent or blank ("")
#   - coerces YAML lists to plain vectors (e.g. state_name_filter)
cfg_get <- function(key, default) {
  if (!exists(".CONFIG", envir = .GlobalEnv)) return(default)
  cfg <- get(".CONFIG", envir = .GlobalEnv)
  if (is.null(cfg[[key]])) return(default)
  val <- cfg[[key]]
  if (is.list(val)) val <- unlist(val, use.names = FALSE)
  if (is.character(val) && length(val) == 1 && !nzchar(val)) return(default)  # blank => default
  val
}
